src/client.c

```diff
@@ -735,28 +736,40 @@ int trilogy_escape(trilogy_conn_t *conn, const char *str, size_t len, const char
 
     b->len = 0;
 
-    if (conn->server_status & TRILOGY_SERVER_STATUS_NO_BACKSLASH_ESCAPES) {
-        for (size_t i = 0; i < len; i++) {
-            const uint8_t c = (uint8_t)str[i];
+    // Escaped string will be at least as large as the source string,
+    // so might as well pre-expand the buffer.
+    CHECKED(trilogy_buffer_expand(b, len));
+
+    uint8_t *cursor = (uint8_t *)str;
+    const uint8_t *end = cursor + len;
 
-            if (c == '\'') {
-                CHECKED(trilogy_buffer_putc(b, '\''));
-                CHECKED(trilogy_buffer_putc(b, '\''));
+    if (conn->server_status & TRILOGY_SERVER_STATUS_NO_BACKSLASH_ESCAPES) {
+        while (cursor < end) {
+            uint8_t *next_backslash = memchr(cursor, '\'', (size_t)(end - cursor));
```

Comment:
This isn't actually the next backslash, it's the next quote. Maybe we call this `next_escape`?